linalg/sundials.cpp

```diff
@@ -659,331 +668,556 @@ void ARKODESolver::Step(Vector &x, double &t, double &dt)
+
+   // Wrap linear solver as SUNLinearSolver and SUNMatrix
+   LSM = SUNLinSolNewEmpty();
+   MFEM_VERIFY(LSM, "error in SUNLinSolNewEmpty()");
+
+   LSM->content      = this;
+   LSM->ops->gettype = LSGetType;
+   LSM->ops->solve   = ARKStepSolver::MassSysSolve;
+   LSA->ops->free    = LSFree;
+
+   M = SUNMatNewEmpty();
+   MFEM_VERIFY(M, "error in SUNMatNewEmpty()");
+
+   M->content      = this;
+   M->ops->getid   = SUNMatGetID;
+   M->ops->matvec  = ARKStepSolver::MassMult1;
+   M->ops->destroy = MatDestroy;
+
+   // Attach the linear solver and matrix
+   flag = ARKStepSetMassLinearSolver(sundials_mem, LSM, M, tdep);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetLinearSolver()");
+
+   // Set the linear system function
+   flag = ARKStepSetMassFn(sundials_mem, ARKStepSolver::MassSysSetup);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetMassFn()");
+}
+
+void ARKStepSolver::UseSundialsMassLinearSolver(int tdep)
+{
+   // Free any existing matrix and linear solver
+   if (M != NULL)   { SUNMatDestroy(A); M = NULL; }
+   if (LSM != NULL) { SUNLinSolFree(LSM); LSM = NULL; }
+
+   // Create linear solver
+   LSM = SUNLinSol_SPGMR(y, PREC_NONE, 0);
+   MFEM_VERIFY(LSM, "error in SUNLinSol_SPGMR()");
+
+   // Attach linear solver
+   flag = ARKStepSetMassLinearSolver(sundials_mem, LSM, NULL, tdep);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetMassLinearSolver()");
+
+   // Attach matrix multiplication function
+   flag = ARKStepSetMassTimes(sundials_mem, NULL, ARKStepSolver::MassMult2,
+                              this);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetMassTimes()");
 }
 
+void ARKStepSolver::SetStepMode(int itask)
+{
+   step_mode = itask;
+}
+
+void ARKStepSolver::SetSStolerances(double reltol, double abstol)
+{
+   flag = ARKStepSStolerances(sundials_mem, reltol, abstol);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSStolerances()");
+}
+
+void ARKStepSolver::SetMaxStep(double dt_max)
+{
+   flag = ARKStepSetMaxStep(sundials_mem, dt_max);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetMaxStep()");
+}
+
+void ARKStepSolver::SetOrder(int order)
+{
+   flag = ARKStepSetOrder(sundials_mem, order);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetOrder()");
+}
 
-static inline KINMem Mem(const KinSolver *self)
+void ARKStepSolver::SetERKTableNum(int table_num)
 {
-   return KINMem(self->SundialsMem());
+   flag = ARKStepSetTableNum(sundials_mem, -1, table_num);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetTableNum()");
 }
 
-// static method
-int KinSolver::Mult(const N_Vector u, N_Vector fu, void *user_data)
+void ARKStepSolver::SetIRKTableNum(int table_num)
+{
+   flag = ARKStepSetTableNum(sundials_mem, table_num, -1);
+   MFEM_VERIFY(flag == ARK_SUCCESS, "error in ARKStepSetTableNum()");
+}
+
+void ARKStepSolver::SetIMEXTableNum(int etable_num, int itable_num)
+{
+   flag = ARKStepSetTableNum(sundials_mem, itable_num, itable_num);
```

Comment:
Shouldn't one of the parameters use `etable_num`? (I noticed this because my editor shows `etable_num` as unused.)

Reply:
Yes, the second input to `ARKStepSteTableNum` should be `etable_num`. I'll fix this in a sec.